Mathlib/Data/Multiset/Order.lean

```diff
@@ -0,0 +1,552 @@
+      rw [this]
+      have : P + N = N + P := by apply add_comm P N
+      simp_all only [ge_iff_le]
+    rw [this]
+    have : M + P - N + N = M + P := by
+      have : M + P - N + N = (M + P) ∪ N := by apply Eq.refl
+      have : (M + P) ∪ N = M + P:= by
+        apply Multiset.eq_union_left
+        have : M ≤ M + P := by simp_all only [ge_iff_le, le_add_iff_nonneg_right, zero_le]
+        apply le_trans h this
+      simp_all only [ge_iff_le]
+    rw [this]
+    have : M - N + N = M := by
+      have : M = M - N + N := by
+        apply mem_leq_diff
+        exact h
+      rw [← this]
+    simp_all only [ge_iff_le]
+  simp_all only [ge_iff_le, add_left_inj]
+
+lemma le_eq_sub : ∀ (M N P Q : ℕ) , M ≤ P → M + N = P + Q → N = Q + (P - M):= by
+  intros M N P Q h0 h1
+  have := tsub_add_cancel_of_le h0
+  linarith
+
+lemma double_split {α} [dec : DecidableEq α]:
+    ∀ (M N P Q: Multiset α) ,  M + N = P + Q → N = N ∩ Q + (P - M)  := by
+  intros M N P Q h
+  ext x
+  rw [Multiset.count_add]
+  rw [Multiset.count_inter]
+  rw [Multiset.count_sub]
+  have H0 : M.count x + N.count x = P.count x + Q.count x := by
+    rw [Multiset.ext] at h
+    simp_all only [Multiset.mem_add, Multiset.count_add]
+  if l_u : M.count x ≤ P.count x then
+    have : N.count x ≥ Q.count x := by linarith
+    simp_all only [ge_iff_le, min_eq_right]
+    apply le_eq_sub (M.count x) (N.count x) (P.count x) (Q.count x)
+    · simp_all
+    · exact H0
+  else
+    simp_all only [not_le, gt_iff_lt]
+    have : Multiset.count x N ≤ Multiset.count x Q := by linarith
+    have:= le_of_lt l_u
+    simp_all
+
+lemma in_notin_diff {α} [DecidableEq α]:
+    ∀ (x : α) (X Y: Multiset α) ,  x ∈ X → x ∉ Y → x ∈ X - Y  := by
+  intros x X Y x_in_X x_notin_Y
+  have : Multiset.count x X ≥ 1 := by
+    rw [← Multiset.one_le_count_iff_mem] at x_in_X
+    exact x_in_X
+  rw [← Multiset.one_le_count_iff_mem]
+  rw [Multiset.count_sub]
+  simp_all only [not_false_eq_true, Multiset.count_eq_zero_of_not_mem, tsub_zero]
+
+-- `MultisetLT` is transitive. Two lemmas needed: double_split, in_notin_diff
+lemma LT_trans {α} [pre : Preorder α] [dec : DecidableEq α]:
+    ∀ (M N P : Multiset α) , MultisetLT N M → MultisetLT P N → MultisetLT P M := by
+  intros M N P LTNM LTPN
+  rcases LTNM with ⟨Y1, X1, Z1, X1_ne, N1_def, M1_def, Ord1⟩
+  rcases LTPN with ⟨Y2, X2, Z2, _, P2_def, N2_def, Ord2⟩
+  apply MultisetLT.MLT (Y2 + (Y1 - X2)) (X1 + (X2 - Y1)) (Z1 ∩ Z2)
+  · simp_all only [Multiset.empty_eq_zero, ne_eq, add_eq_zero_iff, false_and, not_false_eq_true]
+  · rw [P2_def]
+    have : Z1 ∩ Z2 + (Y2 + (Y1 - X2)) = Z1 ∩ Z2 + (Y1 - X2) + Y2 := by
+      have : (Y2 + (Y1 - X2)) = (Y1 - X2) + Y2 := by rw [add_comm]
+      rw [this]
+      rw [add_assoc]
+    rw [this]
+    have : Z1 ∩ Z2 + (Y1 - X2) = Z2 ∩ Z1 + (Y1 - X2) := by
+      rw [Multiset.inter_comm]
+    rw [this]
+    rw [← double_split]
+    rw [add_comm]
+    rw [← N2_def]
+    rw [N1_def]
```

Comment:
```suggestion
    rw [this]
    rw [← double_split]
    rw [add_comm]
    rw [← N2_def]
    rw [N1_def]
```
Please put all rewrites in a single `rw` whenever possible. This comment applies in other places in the PR too.

Reply:
Thanks! One thing I notice is that I can only combine them into two lines but not one and I thought they should be equivalent.

Reply:
This is interesting! If I had to guess what is going on here: it is probably because `rw` tries `rfl` after the end of the rewrites, and in this case, there are 2 goals, the first goal is being closed by the `rfl` which is called at the end of the first `rw`. This can be verified by replacing the first `rw` with `rewrite`, which rewrites without trying `rfl` at the end. If an error occurred, then my guess above is correct.

Reply:
Thanks! I think you are right :-)

This: 
`
rw [this,← double_split]
rw [add_comm, ← N2_def, N1_def]
`
works while

`
rewrite [this,← double_split]
rw [add_comm, ← N2_def, N1_def]
`
gives an error.